### PR TITLE
Irreducible loops cleanup

### PR DESCRIPTION
**Pull request by ailrst**
Addresses https://github.com/UQ-PAC/BASIL/issues/194 and https://github.com/UQ-PAC/BASIL/issues/194

Refactors the analysis to be iterative rather than recursive, work over blocks rather than Commands, adds flow conditions and a cli flag to disable the analysis/transform. 

It also factors out the algorithm's state from class members + mutable collections into an immutable case class, which I think makes the high level api a bit clearer, but the algorithm itself is probably more difficult to understand.

### CODE REVIEW

File: src/main/scala/analysis/IrreducibleLoops.scala [hunk 1]
```diff
@@ -55,47 +51,138 @@ class Loop(val header: CFGPosition) {
...
-    val funcEntries = cfg.procedures
-    funcEntries.foreach { funcEntry => traverse_loops_dfs(funcEntry, 1) }
-    loops.values.toSet
+  def identify_loops(cfg: Program): State = {
+    cfg.procedures.toSet.map(_.entryBlock).flatten
+      .foldLeft(State())((st, eb) => (traverse_loops_dfs(st, eb, 1)._1))
   }
 
+
+  private def processVisitedNodeOutgoingEdge(istate: State, edge: LoopEdge, DFSPpos: Int): State = {
```
Comment (l-kent) [inline src/main/scala/analysis/IrreducibleLoops.scala:90]: DFSPpos is never used, is there a reason it exists?

File: src/main/scala/analysis/IrreducibleLoops.scala [hunk 2]
```diff
@@ -110,167 +197,145 @@ class LoopDetector(cfg: Program) {
...
-                  if (iloopHeaders(from) != ih) {
-                    thisLoop.reducible = false
-                    thisLoop.reentries += edge
-                  }
-                } else {
-                  // `from` must be outside the loop
-                  thisLoop.reducible = false
-                  thisLoop.reentries += edge
-                }
+  private def traverse_loops_dfs(_istate: State, _b0: Block, _DFSPpos: Int): (State, Option[Block]) = {
```
Comment (l-kent) [inline src/main/scala/analysis/IrreducibleLoops.scala:200]: Nothing actually uses the second output parameter, what is it for & what does it mean?

**Comment by l-kent** [comment]
This seems fine overall, I'll merge it in once those unused parameters are explained

File: src/main/scala/analysis/IrreducibleLoops.scala (hunk 1 above)
Comment (ailrst) [reply src/main/scala/analysis/IrreducibleLoops.scala:90]: I think I just missed it in the refactor, its been moved into State.

File: src/main/scala/analysis/IrreducibleLoops.scala (hunk 2 above)
Comment (ailrst) [reply src/main/scala/analysis/IrreducibleLoops.scala:200]: Do you mean the `Option[Block]`? I believe this is the first loop header following _b0, this is a holdover from the recursive style of the algorithm which traverses all the successors and tags block that is returned as a header. It is retained in the refactor so the psuedo-return  `retval` that is set in the `FinishProcessNode` state in the explicit call stack form matches the return value of the whole function.

Comment (l-kent) [reply src/main/scala/analysis/IrreducibleLoops.scala:200]: The only time traverse_loops_dfs gets called, it has the tuple access ._1 at the end which only returns the State. Do we need to have this second output parameter at all? If there is some reason we want to keep it, there should be a comment documenting what it means.

Comment (ailrst) [reply src/main/scala/analysis/IrreducibleLoops.scala]: I've just removed it as it isn't really necessary.

**Pull request merged**